### PR TITLE
GitHub Actions CI with micromamba

### PR DESCRIPTION
**Pull request by jGaboardi**
Address in #613 

This PR refactors CI (`unittests.yml`) to utilize [`micromamba`](https://github.com/mamba-org/provision-with-micromamba). Total CI runtime is reduced from [~52 minutes](https://github.com/pysal/spaghetti/actions/runs/781577584) to [~22 minutes](https://github.com/pysal/spaghetti/actions/runs/783416495) to run the full suite of 46 jobs.

### CODE REVIEW

**Review by martinfleis — APPROVED** [review]
This looks fine to me. Just out of curiosity, why we have such an extensive test suite with 48 different combinations? Is there a history of breaking only specific OS/installations? Just for a reference, geopandas uses 12 combinations and we never had an issue that something was not captured.

**Comment by martinfleis** [comment]
Just realised that another option how to go around the not yet supported `pip` section of YAML can be a section installing it manually within the action. Instead of getting them from the env file. That should work within a micromamba environment.

**Comment by jGaboardi** [comment]
> This looks fine to me. Just out of curiosity, why we have such an extensive test suite with 48 different combinations? Is there a history of breaking only specific OS/installations? Just for a reference, geopandas uses 12 combinations and we never had an issue that something was not captured.

* OS
  * Yes, though I can't locate the specific issue(s) now. At several times in the past (when we were using only Travis) things weren't passing specifically on Windows, so we decided to implement Appveyor then soon after migrated everything to GHA. 
* base vs. `PLUS`
  * A nice chunk of `spaghetti`'s functionality is optional (for the time being) and requires `geopandas` to run, though the basic functionality does not. Although running tests for `PLUS` is a superset of the base functionality, I prefer to run tests also on just the base functionality to catch any weird problem before they arise for users, etc. It does take longer, but I see as insurance against rare edge cases that may occur.
* `PYPI` vs. `GIT` 
  * We are doing two sets of tests with two versions of `libpysal`: the `PYPI` (published release) and `GIT` (bleeding edge), though now this terminology seems confusing. Maybe renaming to `STABLE` and `DEV` might be a better description?
* Python
  * In the very near future I am planning on dropping support Python 3.6, which will eliminate a chunk of tests.

**Comment by jGaboardi** [comment]
> Just realised that another option how to go around the not yet supported `pip` section of YAML can be a section installing it manually within the action. Instead of getting them from the env file. That should work within a micromamba environment.

This is a nice catch. Let's try this out.

**Comment by martinfleis** [comment]
@jGaboardi I understand the options but if you look at geopandas CI we cover the same cases. The point is that if you have an env with master libpysal, do you need to test it on all OS's with all versions of Python? I'd say no, the potential bugs will be the same on all (in 99% of cases). I don't mind having all these it just feels a bit unnecessary to me esp. since the current version runs for an hour.

I normally try not to use services that are free to OSS excessively (e.g. that is why we don't build docs for PRs in geopandas) but since GH is owned by Microsoft that probably doesn't matter much here :D.

**Comment by jGaboardi** [comment]
> @jGaboardi I understand the options but if you look at geopandas CI we cover the same cases. The point is that if you have an env with master libpysal, do you need to test it on all OS's with all versions of Python? I'd say no, the potential bugs will be the same on all (in 99% of cases). I don't mind having all these it just feels a bit unnecessary to me esp. since the current version runs for an hour.
> 
> I normally try not to use services that are free to OSS excessively (e.g. that is why we don't build docs for PRs in geopandas) but since GH is owned by Microsoft that probably doesn't matter much here :D.

Maybe I need to rethink this anyway... and I am open to better ways of doing things. My two priorities are (1) getting tests run as quickly as possible while covering as many cases as possible, and (2) Not overusing services [but more for environmental footprint than cost to Microsoft 🙃].

Let's get this current situation worked out with `micromamba` then start a new PR for pruning down tests? I would like more of your input on this.

**Comment by martinfleis** [comment]
> Let's get this current situation worked out with micromamba then start a new PR for pruning down tests?

Sounds good

**Comment by jGaboardi** [comment]
@martinfleis How does this look?

File: .github/workflows/unittests.yml [hunk 1]
```diff
@@ -10,45 +10,54 @@
...
-       - run: conda config --show
-       - run: pytest -v spaghetti --cov=spaghetti --doctest-modules --cov-config=.coveragerc --cov-report=xml
-       - name: codecov (${{ matrix.os }}, ${{ matrix.environment-file }})
+           environment-file: ${{ matrix.environment-file }}
+           micromamba-version: 'latest'
+      
+       - name: install bleeding edge libpysal - bash
+         shell: bash -l {0}
+         run: |
+           pip install pip
```
Comment (martinfleis) [inline .github/workflows/unittests.yml:39]: This looks strange to my eyes. What does it do?

Comment (jGaboardi) [reply .github/workflows/unittests.yml:39]: It installs `pip`. Micromamba is a completely base installation, so I thought it's needed. Maybe it's not needed?

Comment (jGaboardi) [reply .github/workflows/unittests.yml:39]: Maybe it's better to put the `pip` requirement back in the `ci/*DEV.yaml` files and the we can cut the `pip install pip` line?

Comment (martinfleis) [reply .github/workflows/unittests.yml:39]: How do you install pip with pip if you don't have pip? :)

> Maybe it's better to put the pip requirement back in the ci/*DEV.yaml

I think so.
